adapters/http-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -1138,32 +1144,23 @@ protected boolean isCommandValid(final Command command, final Span currentSpan)
      * The created timer's ID is put to the routing context using key {@link #KEY_TIMER_ID}.
      *
      * @param ctx The device's currently executing HTTP request.
-     * @param responseReady The future to complete when the time has expired.
+     * @param responseReadyTracker The response ready tracker.
      * @param delaySecs The number of seconds to wait for a command.
-     * @param waitForCommandSpan The span tracking the command reception.
      */
     private void addCommandReceptionTimer(
             final RoutingContext ctx,
-            final AtomicBoolean requestProcessed,
-            final Handler<AsyncResult<Void>> responseReady,
-            final long delaySecs,
-            final Span waitForCommandSpan) {
+            final ResponseReadyTracker responseReadyTracker,
+            final long delaySecs) {
 
         final Long timerId = ctx.vertx().setTimer(delaySecs * 1000L, id -> {
 
             log.trace("time to wait [{}s] for command expired [timer id: {}]", delaySecs, id);
 
-            if (requestProcessed.compareAndSet(false, true)) {
-                // no command to be sent,
-                // send empty response
-                setTtdStatus(ctx, TtdStatus.EXPIRED);
-                waitForCommandSpan.log(String.format("time to wait for command expired (%ds)", delaySecs));
-                waitForCommandSpan.finish();
-                responseReady.handle(Future.succeededFuture());
-            } else {
+            final boolean canceled = responseReadyTracker.cancel(String.format("time to wait for command expired (%ds)", delaySecs), TtdStatus.EXPIRED);
+            if (canceled) {
                 // a command has been sent to the device already
                 log.trace("response already sent, nothing to do ...");
-            }
+            } // otherwise no command to be sent, send empty response
```

Comment:
It seems to me this should be `if (!canceled) {` here.

Reply:
yes. I'll fix it